packages/flutter/lib/src/services/text_formatter.dart

```diff
@@ -370,6 +370,10 @@ class LengthLimitingTextInputFormatter extends TextInputFormatter {
     TextEditingValue oldValue, // unused.
     TextEditingValue newValue,
   ) {
+    // Always return new value when it's composing.
+    if (newValue.isComposingRangeValid) {
```

Comment:
Using `isComposingRangeValid` is a bit too strong. It's mainly used for out of range check. To check whether `newValue` contains composing text, I think `newValue.composing.isValid` is preferred. Yeah I'm sorry the name is a bit confusing.

Reply:
Also applies to the changes below.

Reply:
I think the behavior change is worth documenting. We should probably add a "#Multistage text input" section or "#Composing Text" section somewhere in this class for visibility, and document the fact that the [maxLength] constraint is going to be temporarily lifted when the user starts composing, and when the composing finishes, the text may get truncated so it does not exceed [maxLength].

Reply:
Also, ideally we shouldn't allow the user to start composing when they've already reached the length limit.

Reply:
> Also, ideally we shouldn't allow the user to start composing when they've already reached the length limit.

Strongly agreed with this one! I'll add it.